examples/matching-engine/tests/transaction.rs

```diff
@@ -0,0 +1,263 @@
+}
+
+/// Test creating a matching engine, pushing some orders, canceling some and
+/// seeing how the transactions went.
+///
+/// The operation is done in exactly the same way with the same amounts
+/// and quantities as the corresponding end to end test.
+#[tokio::test]
+async fn single_transaction() {
+    let (validator, bytecode_id) = TestValidator::with_current_bytecode().await;
+
+    let mut fungible_chain_a = validator.new_chain().await;
+    let owner_a = AccountOwner::from(fungible_chain_a.public_key());
+    let mut fungible_chain_b = validator.new_chain().await;
+    let owner_b = AccountOwner::from(fungible_chain_b.public_key());
+    let mut matching_chain = validator.new_chain().await;
+    let admin_account = AccountOwner::from(matching_chain.public_key());
+
+    let fungible_bytecode_id_a = fungible_chain_a.publish_bytecodes_in("../fungible").await;
+    let fungible_bytecode_id_b = fungible_chain_b.publish_bytecodes_in("../fungible").await;
+
+    let initial_state_a =
+        InitialStateBuilder::default().with_account(owner_a, Amount::from_tokens(10));
+    let params_a = fungible::Parameters::new("A");
+    let token_id_a = fungible_chain_a
+        .create_application::<FungibleTokenAbi>(
+            fungible_bytecode_id_a,
+            params_a,
+            initial_state_a.build(),
+            vec![],
+        )
+        .await;
+    let initial_state_b =
+        InitialStateBuilder::default().with_account(owner_b, Amount::from_tokens(9));
+    let params_b = fungible::Parameters::new("B");
+    let token_id_b = fungible_chain_b
+        .create_application::<FungibleTokenAbi>(
+            fungible_bytecode_id_b,
+            params_b,
+            initial_state_b.build(),
+            vec![],
+        )
+        .await;
+
+    fungible_chain_a.register_application(token_id_b).await;
+    fungible_chain_b.register_application(token_id_a).await;
+
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, Some(Amount::from_tokens(10))),
+        (owner_b, None),
+    ] {
+        let value = FungibleTokenAbi::query_account(token_id_a, &fungible_chain_a, owner).await;
+        assert_eq!(value, amount);
+    }
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, None),
+        (owner_b, Some(Amount::from_tokens(9))),
+    ] {
+        let value = FungibleTokenAbi::query_account(token_id_b, &fungible_chain_b, owner).await;
+        assert_eq!(value, amount);
+    }
+
+    let tokens = [token_id_a, token_id_b];
+    let matching_parameter = Parameters { tokens };
+    let matching_id = matching_chain
+        .create_application::<MatchingEngineAbi>(
+            bytecode_id,
+            matching_parameter,
+            (),
+            vec![token_id_a.forget_abi(), token_id_b.forget_abi()],
+        )
+        .await;
+    // Doing the registrations
+    matching_chain.register_application(token_id_a).await;
+    matching_chain.register_application(token_id_b).await;
```

Comment:
I'm surprised `create_application` works _before_ this, even though those are required.

Reply:
I agree, this is strange. But yes it does.